message.go

```diff
@@ -790,6 +881,14 @@ func (m *Message) Marshal() ([]byte, error) {
 	if nsegs == 0 {
 		return nil, newError("marshal: message has no segments")
 	}
+	if nsegs == 1 {
+		if psa, ok := m.Arena.(*pooledSegmentArena); ok {
+			// fast path for single pooled segment
+			b := *psa
+			binary.LittleEndian.PutUint32(b[4:], uint32(len(m.firstSeg.data)/int(wordSize)))
+			return b[:hdrSize+len(m.firstSeg.data)], nil
```

Comment:
I really don't like special-casing this in `Marshal`, for a couple reasons, in decreasing order of importance:

- This actually changes the semantics: If the arena is a `pooledSegmentArena`, then the underlying buffer is *shared* with the message, so further mutations will also affect the marshaled value, whereas this is not the case with the other arena types.
- I worry this is introducing a performance footgun; switching arena types may silently introduce a copy that wasn't there before.
- Tying this to pool allocation feels like a violation of separation of concerns; it seems like we should in principle be able to get this to work with `SingleSegment(nil)` too.

Perhaps instead, we could make this a method on the arena itself. We could then just document that `Message.Marshal` makes a full copy and point users to the alternatives.

We would have to expose the new method somehow. Some options:

1. Export the type as I suggested elsewhere.
2. Maybe add a new method to arena that returns a `[][]byte` which is zero copy, but notably is *not* returning a list of segments, but the serialized form of the message, with some arbitrary division. Code that needs a `[]byte` could check the length to see if it got lucky (which it would for pooled arenas, and we could optimize this for SingleSegment if the  user passes in nil).
3. Add a new interface which embeds arena, that *some* arenas can implement, but not all need to.

Note that (1) is not mutually exclusive with the other options.

If we go with (3) we have a choice of making the method signature as described in `(2)` or just returning `[]byte`; I can see arguments both ways.

(2) is technically backwards-incompatible, but it's a small change and we'll probably make bigger breaking changes before we tag v3, so I don't consider that a dealbreaker.

Reply:
Thanks, I thought it would be simpler to move all the serialization logic to a single function.